kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/client/DefaultParquetHandler.java

```diff
@@ -63,23 +63,24 @@ public void close() throws IOException {
 
             @Override
             public boolean hasNext() {
-                // There is no file in reading or the current file being read has no more data
-                // initialize the next file reader or return false if there are no more files to
-                // read.
-                if (currentFileReader == null || !currentFileReader.hasNext()) {
+                if (currentFileReader != null && currentFileReader.hasNext()) {
+                    return true;
+                } else {
+                    // There is no file in reading or the current file being read has no more data.
+                    // Initialize the next file reader or return false if there are no more files to
+                    // read.
                     Utils.closeCloseables(currentFileReader);
                     currentFileReader = null;
                     if (fileIter.hasNext()) {
                         currentFile = fileIter.next();
                         FileStatus fileStatus =
                             InternalScanFileUtils.getAddFileStatus(currentFile.getScanFileRow());
                         currentFileReader = batchReader.read(fileStatus.getPath(), physicalSchema);
+                        return hasNext(); // recurse since it's possible the loaded file is empty
```

Comment:
Can you check if this bug exists for DefaultJsonHandler too?

Reply:
Also the `LogReplay`.

Reply:
https://github.com/delta-io/delta/issues/2251